django/db/models/sql/query.py

```diff
@@ -1336,9 +1336,15 @@ def build_filter(self, filter_expr, branch_negated=False, current_negated=False,
                 # (col IS NULL OR col != someval)
                 #   <=>
                 # NOT (col IS NOT NULL AND col = someval).
+                # If someval refers to another field on the model,
+                # someval IS NOT NULL will be added as well:
+                # NOT (col IS NOT NULL AND someval IS NOT NULL
+                #     AND col = someval)
                 lookup_class = targets[0].get_lookup('isnull')
                 col = self._get_col(targets[0], join_info.targets[0], alias)
                 clause.add(lookup_class(col, False), AND)
+                if isinstance(value, Col):
```

Comment:
This should be added only if column is nullable:
```python
if isinstance(value, Col) and self.is_nullable(value.target):
    clause.add(lookup_class(value, False), AND)
```

Reply:
good catch, thanks!